AsyncNats/NatsServerPool.cs

```diff
@@ -0,0 +1,176 @@
+                            .Where(s => s != _lastSelectedServer)
+                            .OrderBy(s => _random.Next())
+                            .First();
+                    }
+                    else
+                    {
+                        if(_lastSelectedServer != null)
+                        {
+                            //return server list in order
+                            if (combinedServers.IndexOf(_lastSelectedServer) + 1 >= combinedServers.Count)
+                                selectedServer = combinedServers[0];
+                            else
+                                selectedServer = combinedServers[combinedServers.IndexOf(_lastSelectedServer) + 1];
+                        }
+                    }
+                }
+
+                _lastSelectedServer = selectedServer;
+
+                return selectedServer!;
+            }            
+        }
+
+        public void AddDiscoveredServers(IEnumerable<string> servers)
+        {
+            if (servers == null || servers.Count()==0) return;
+
+            if (!_options.ServersOptions.HasFlag(NatsServerPoolFlags.AllowDiscovery))
+                return;
+
+            lock (_sync)
+            {
+                _discoveredServers = new List<NatsHost>();
+
+                foreach (var server in servers.OrderBy(s=>Guid.NewGuid())) //randomize discovered
+                    AddServer(_discoveredServers, server);
+            }
+            
+        }
+
+        private NatsHost ParseAndNormalizeServerEntry(string server)
+        {
+            var schemaIndex = server.IndexOf("://");
+            if (schemaIndex > 0)
+            {
+                server = server.Substring(schemaIndex + 3);
+            }
+
+            if (!server.Contains(':'))
+            {
+                server = $"{server}:4222"; //default nats port
+            }
+
+            var split = server.Split(':');
+
+            if (split.Length != 2)
+                throw new FormatException($"invalid server string {server}");
+
+            var stringHost = split[0];
+            var stringPort = split[1];
+
+            if(!int.TryParse(stringPort, out int port))
+                throw new FormatException($"invalid server string {server}");
+
+            var checkHostNameResult = Uri.CheckHostName(stringHost);
+            if(checkHostNameResult==UriHostNameType.Unknown)
+                throw new FormatException($"invalid server string {server}");
+
+            return new NatsHost(stringHost,port);
+            
+        }
+
+        private void AddServer(List<NatsHost> list, string server)
+        {
+            lock (_sync)
+            {
+                var parsed = ParseAndNormalizeServerEntry(server);
+
+                if (!list.Contains(parsed))
+                {
+                    list.Add(parsed);
+                }
+            }
+        }               
+
+        internal class NatsHost:IEquatable<NatsHost>

```

Comment:
How is NatsHost different vs DnsEndPoint:
https://docs.microsoft.com/en-us/dotnet/api/system.net.dnsendpoint?view=net-6.0

Reply:
You are right, I had some extra property there and that was later removed and it ended up being the same representation.